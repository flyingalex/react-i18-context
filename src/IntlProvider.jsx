import React from 'react';
import { PropTypes } from 'prop-types';

const IntlContext = React.createContext();

// put this at the top of the app
class IntlProvider extends React.Component {
  constructor(props) {
    super(props);
    const { languages, locale } = this.props;
    if (locale === undefined) {
      throw new Error(`no initial lang[${locale}], please set locale variable`);
    }

    if (!Object.keys(languages).includes(locale)) {
      throw new Error(`no such language[${locale}]`);
    }
    this.state = {
      locale: languages[locale],
      langs: Object.keys(languages),
    };
  }

  changeLan = locale => {
    const { langs } = this.state;
    const { languages } = this.props;
    if (langs.includes(locale)) {
      this.setState({ locale: languages[locale] });
    } else {
      throw new Error(`no such language[${locale}]`);
    }
  };

  render() {
    const { locale, langs } = this.state;
    const { children } = this.props;
    return (
      <IntlContext.Provider
        value={{ langs, locale, changeLan: this.changeLan }}
      >
        {children}
      </IntlContext.Provider>
    );
  }
}

IntlProvider.propTypes = {
  children: PropTypes.object,
  languages: PropTypes.object,
  init: PropTypes.string,
};

const FormatMsg = props => {
  const { id, inject = {} } = props;
  return (
    <IntlContext.Consumer>
      {intl => {
        let text = intl.locale[id];
        const injectKeys = Object.keys(inject);
        if (text === undefined) {
          throw new Error(`no such id(${id})`);
        }
        text = text.replace(/{(.*?)}/g, matched => {
          const key = matched.replace('{', '').replace('}', '');
          if (inject[key]) {
            return inject[key];
          }
          return matched;
        });
        return text;
      }}
    </IntlContext.Consumer>
  );
};

const DefineLangue = props => {
  return (
    <IntlContext.Consumer>
      {intl => (
        <div onClick={() => intl.changeLan(props.locale)}>{props.children}</div>
      )}
    </IntlContext.Consumer>
  );
};

const InjectIntlLangWrapper = Component => {
  const { name } = Component.prototype.constructor;
  class Consumer extends React.Component {
    render() {
      return (
        <IntlContext.Consumer>
          {intl => <Component changeLan={intl.changeLan} />}
        </IntlContext.Consumer>
      );
    }
  }
  Consumer.displayName = `IntlWrapper${name}`;
  return Consumer;
};

// apis
export default IntlProvider;
export { FormatMsg, DefineLangue, InjectIntlLangWrapper };
