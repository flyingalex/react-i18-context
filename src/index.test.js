import React from 'react';
import IntlProvider, {
  FormatMsg,
  LocaleSet,
  InjectIntlLangWrapper,
} from './index';
import renderer from 'react-test-renderer';

const en = {
  test: 'hello {name}',
};
const zh = {
  test: '你好',
};

const messages = {
  en,
  zh,
};

const TestComponent = props => (
  <div style={{ textAlign: 'center', marginTop: '50px' }} />
);

const HOCApp = Component =>
  renderer.create(
    <IntlProvider messages={messages} locale="en">
      <Component />
    </IntlProvider>,
  );

test('IntlProvider config', () => {
  const component = HOCApp(TestComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('InjectIntlLangWrapper inject', () => {
  const InjectIntlLangWrapperComponent = InjectIntlLangWrapper(TestComponent);
  const component = HOCApp(InjectIntlLangWrapperComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FormatMsg', () => {
  const FormatMsgComponent = () => (
    <FormatMsg id="test" inject={{ name: 'inject' }} />
  );
  const component = HOCApp(FormatMsgComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LocaleSet', () => {
  const LocaleSetComponent = () => (
    <LocaleSet locale="en">
      <button>"en"</button>
    </LocaleSet>
  );
  const component = HOCApp(LocaleSetComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
