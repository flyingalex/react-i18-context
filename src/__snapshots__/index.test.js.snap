// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefineLangue 1`] = `
<div
  onClick={[Function]}
>
  <button>
    "en"
  </button>
</div>
`;

exports[`FormatMsg 1`] = `"hello inject"`;

exports[`InjectIntlLangWrapper inject 1`] = `
<div
  style={
    Object {
      "marginTop": "50px",
      "textAlign": "center",
    }
  }
/>
`;

exports[`IntlProvider config 1`] = `
<div
  style={
    Object {
      "marginTop": "50px",
      "textAlign": "center",
    }
  }
/>
`;
